File: python/sparkbricks/src/sparkbricks/connect.py
"""Spark session connection management with multi-auth support.

Provides get_spark() for creating/retrieving a cached DatabricksSession,
and close_spark() for cleanup. Supports both OAuth and PAT authentication.
"""

from __future__ import annotations

import os
from typing import TYPE_CHECKING

from sparkbricks.auth import AuthType, _ensure_cli_path, get_workspace_client

if TYPE_CHECKING:
    from pyspark.sql import SparkSession

# Module-level spark session (lazy initialized)
_spark: "SparkSession | None" = None
_env_loaded: bool = False


def _load_dotenv() -> None:
    """Load .env file if python-dotenv is available.

    Automatically searches current directory and parents for .env file.
    Only loads once per session.
    """
    global _env_loaded
    if _env_loaded:
        return

    try:
        from dotenv import load_dotenv

        load_dotenv()  # Searches current dir and parents
        _env_loaded = True
    except ImportError:
        pass  # dotenv not installed, skip silently


def get_spark(
    host: str | None = None,
    cluster_id: str | None = None,
    token: str | None = None,
    profile: str = "DEFAULT",
    auth_type: str = "auto",
    auto_start_cluster: bool = True,
) -> "SparkSession | None":
    """Get or create a Databricks Spark session with multi-auth support.

    Supports both OAuth (via Databricks CLI) and PAT authentication.
    Automatically loads configuration from .env file (if python-dotenv installed).
    Automatically starts the cluster if it's not running.

    Authentication priority (when auth_type="auto"):
    1. If token parameter provided -> PAT auth
    2. If DATABRICKS_TOKEN env var set -> PAT auth
    3. Otherwise -> OAuth via Databricks CLI profile

    Args:
        host: Databricks workspace URL (or set DATABRICKS_HOST in .env)
        cluster_id: Cluster ID (or set DATABRICKS_CLUSTER_ID in .env)
        token: Personal Access Token (or set DATABRICKS_TOKEN in .env)
        profile: Databricks CLI profile name (for OAuth)
        auth_type: "auto", "oauth", or "pat"
        auto_start_cluster: If True, start cluster if not running

    Returns:
        SparkSession or None if connection fails

    Example:
        >>> # Auto-detect auth (PAT if DATABRICKS_TOKEN set, else OAuth)
        >>> spark = get_spark()
        >>> df = spark.sql("SELECT 1")

        >>> # Explicit PAT auth
        >>> spark = get_spark(token="dapi...")

        >>> # Explicit OAuth with specific profile
        >>> spark = get_spark(auth_type="oauth", profile="work")
    """
    global _spark

    if _spark is not None:
        return _spark

    # Load .env file if available (python-dotenv)
    _load_dotenv()

    # Import here to avoid circular imports
    from sparkbricks.cluster import ensure_cluster_running

    # Determine effective auth type
    effective_token = token or os.environ.get("DATABRICKS_TOKEN")
    effective_auth_type = auth_type
    if auth_type == "auto":
        effective_auth_type = "pat" if effective_token else "oauth"

    # Setup environment based on auth type
    if effective_auth_type == "pat":
        # PAT mode - set environment directly
        effective_host = host or os.environ.get("DATABRICKS_HOST")
        if not effective_host:
            print("ERROR: DATABRICKS_HOST required when using token authentication")
            return None
        if not effective_token:
            print("ERROR: Token required for PAT authentication")
            return None

        os.environ["DATABRICKS_HOST"] = effective_host
        os.environ["DATABRICKS_TOKEN"] = effective_token
        # Remove profile-based auth when using PAT
        os.environ.pop("DATABRICKS_CONFIG_PROFILE", None)
    else:
        # OAuth mode - ensure CLI is available
        _ensure_cli_path()
        if profile:
            os.environ["DATABRICKS_CONFIG_PROFILE"] = profile

    # Set environment if provided
    if host:
        os.environ["DATABRICKS_HOST"] = host
    if cluster_id:
        os.environ["DATABRICKS_CLUSTER_ID"] = cluster_id

    # Get effective cluster_id (from param or env)
    effective_cluster_id = cluster_id or os.environ.get("DATABRICKS_CLUSTER_ID")

    # Ensure cluster is running before connecting
    if auto_start_cluster and effective_cluster_id:
        if not ensure_cluster_running(
            cluster_id=effective_cluster_id,
            host=host or os.environ.get("DATABRICKS_HOST"),
            token=effective_token,
            profile=profile,
            auth_type=effective_auth_type,
        ):
            print("Failed to ensure cluster is running")
            return None

    try:
        from databricks.connect import DatabricksSession

        _spark = DatabricksSession.builder.getOrCreate()
        print(f"Spark connected (version {_spark.version})")
        return _spark
    except ImportError:
        print("databricks-connect not installed")
        print("Install with: uv pip install databricks-connect")
        return None
    except Exception as e:
        print(f"Databricks connection failed: {e}")
        if effective_auth_type == "oauth":
            print("Ensure 'databricks auth login' has been run")
        return None


def close_spark() -> None:
    """Close the cached Spark session and clear the module-level cache.

    Call this when you're done with the Spark session to properly clean up
    resources. This is especially important when called from R via reticulate
    to avoid orphaned Python sessions.

    Example:
        >>> spark = get_spark()
        >>> # ... do work ...
        >>> close_spark()  # Clean up
    """
    global _spark
    if _spark is not None:
        try:
            _spark.stop()
        except Exception as e:
            print(f"Warning: Error stopping Spark session: {e}")
        _spark = None
        print("Spark session closed")

File: r/spaRkbricks/R/connect.R
#' Connect to Databricks Spark Cluster
#'
#' Establishes a connection to a Databricks Spark cluster using either
#' OAuth (via Databricks CLI) or Personal Access Token (PAT) authentication.
#'
#' @param host Databricks workspace URL. If NULL, reads from DATABRICKS_HOST env var.
#' @param cluster_id Databricks cluster ID. If NULL, reads from DATABRICKS_CLUSTER_ID env var.
#' @param token Personal Access Token. If NULL, reads from DATABRICKS_TOKEN env var.
#'              If provided (or env var set), uses PAT auth. Otherwise uses OAuth.
#' @param profile Databricks CLI profile name (for OAuth auth, default: "DEFAULT")
#' @param auth_type Authentication type: "auto" (default), "oauth", or "pat".
#'                  "auto" uses PAT if token is available, otherwise OAuth.
#' @param auto_start If TRUE, automatically start the cluster if not running
#' @param envname Path to Python virtual environment (default: auto-detect)
#'
#' @return A sparklyr connection object
#' @export
#'
#' @examples
#' \dontrun{
#' # Auto-detect auth (PAT if DATABRICKS_TOKEN set, else OAuth)
#' sc <- spark_connect_databricks()
#'
#' # Explicit PAT auth
#' sc <- spark_connect_databricks(token = "dapi...")
#'
#' # Explicit OAuth with specific profile
#' sc <- spark_connect_databricks(auth_type = "oauth", profile = "work")
#'
#' # Use with dplyr
#' library(dplyr)
#' tbl(sc, "catalog.schema.table") |> filter(x > 10) |> collect()
#'
#' # Disconnect properly
#' spark_disconnect_databricks(sc)
#' }
spark_connect_databricks <- function(
    host = NULL,
    cluster_id = NULL,
    token = NULL,
    profile = "DEFAULT",
    auth_type = c("auto", "oauth", "pat"),
    auto_start = TRUE,
    envname = NULL
) {
  auth_type <- match.arg(auth_type)

  # Initialize Python environment using centralized init
  .init_python(envname)

  # Import sparkbricks
  sparkbricks <- reticulate::import("sparkbricks")

  # Get connection parameters from env vars if not provided
  if (is.null(host)) {
    host <- Sys.getenv("DATABRICKS_HOST", unset = NA)
    if (is.na(host)) {
      stop("DATABRICKS_HOST environment variable not set. Set it or pass host parameter.")
    }
  }

  if (is.null(cluster_id)) {
    cluster_id <- Sys.getenv("DATABRICKS_CLUSTER_ID", unset = NA)
    if (is.na(cluster_id)) {
      stop("DATABRICKS_CLUSTER_ID environment variable not set. Set it or pass cluster_id parameter.")
    }
  }

  # Resolve token
  effective_token <- token
  if (is.null(effective_token)) {
    env_token <- Sys.getenv("DATABRICKS_TOKEN", unset = "")
    if (nzchar(env_token)) {
      effective_token <- env_token
    }
  }

  # Determine effective auth type
  if (auth_type == "auto") {
    effective_auth <- if (!is.null(effective_token)) "pat" else "oauth"
  } else {
    effective_auth <- auth_type
  }

  # Validate PAT auth has token
  if (effective_auth == "pat" && is.null(effective_token)) {
    stop("PAT auth requires token parameter or DATABRICKS_TOKEN env var")
  }

  message(sprintf("Connecting to %s (auth: %s)...", host, effective_auth))

  # Start cluster if requested and needed
  if (auto_start) {
    status <- sparkbricks$cluster_status(
      cluster_id = cluster_id,
      host = host,
      token = effective_token,
      profile = profile,
      auth_type = effective_auth
    )
    if (status != "RUNNING") {
      message(sprintf("Cluster status: %s. Starting...", status))
      sparkbricks$start_cluster(
        cluster_id = cluster_id,
        host = host,
        token = effective_token,
        profile = profile,
        auth_type = effective_auth,
        wait = TRUE
      )
    }
  }

  # Create Python Spark session
  spark <- sparkbricks$get_spark(
    host = host,
    cluster_id = cluster_id,
    token = effective_token,
    profile = profile,
    auth_type = effective_auth
  )

  # Set up environment for sparklyr
  if (effective_auth == "pat") {
    # PAT auth - set token directly
    Sys.setenv(DATABRICKS_TOKEN = effective_token)
    Sys.setenv(DATABRICKS_HOST = host)
    Sys.setenv(DATABRICKS_CLUSTER_ID = cluster_id)
  } else {
    # OAuth - extract token from SDK for sparklyr
    py_code <- sprintf("
import os
from databricks.sdk.core import Config

config = Config(
    host='%s',
    cluster_id='%s',
    auth_type='databricks-cli',
    profile='%s'
)

headers = config.authenticate()
token = headers.get('Authorization', '').replace('Bearer ', '')

os.environ['DATABRICKS_TOKEN'] = token
os.environ['DATABRICKS_HOST'] = '%s'
os.environ['DATABRICKS_CLUSTER_ID'] = '%s'
", host, cluster_id, profile, host, cluster_id)

    reticulate::py_run_string(py_code)
  }

  # Connect via sparklyr
  sc <- sparklyr::spark_connect(
    method = "databricks_connect",
    cluster_id = cluster_id
  )

  message(sprintf("Connected to Databricks cluster %s", cluster_id))

  return(sc)
}


#' Disconnect from Databricks Spark Cluster
#'
#' Properly disconnects from the Databricks Spark cluster, closing both the
#' sparklyr connection and the underlying Python DatabricksSession.
#'
#' This function is important because without it, the Python session would
#' remain open (orphaned) after the R session disconnects.
#'
#' @param sc A sparklyr connection object returned by spark_connect_databricks()
#'
#' @return NULL (invisibly)
#' @export
#'
#' @examples
#' \dontrun{
#' sc <- spark_connect_databricks()
#' # ... do work ...
#' spark_disconnect_databricks(sc)
#' }
spark_disconnect_databricks <- function(sc) {
  # Close sparklyr connection first
  sparklyr::spark_disconnect(sc)

  # Close the Python session to avoid orphaned sessions
  tryCatch({
    sparkbricks <- reticulate::import("sparkbricks")
    sparkbricks$close_spark()
    message("Disconnected from Databricks (both R and Python sessions closed)")
  }, error = function(e) {
    message("Note: Could not close Python session. It may already be closed.")
  })

  invisible(NULL)
}
